gadget/ondisk.go

```diff
@@ -130,21 +126,31 @@ func fromSfdiskPartitionType(st string, sfdiskLabel string) (string, error) {
 	}
 }
 
-func blockDeviceSizeInSectors(devpath string) (quantity.Size, error) {
-	// the size is reported in 512-byte sectors
-	// XXX: consider using /sys/block/<dev>/size directly
-	out, err := exec.Command("blockdev", "--getsz", devpath).CombinedOutput()
+func blockdevCmd(cmd, devpath string) (quantity.Size, error) {
+	out, err := exec.Command("blockdev", cmd, devpath).CombinedOutput()
 	if err != nil {
 		return 0, osutil.OutputErr(out, err)
 	}
 	nospace := strings.TrimSpace(string(out))
 	sz, err := strconv.Atoi(nospace)
 	if err != nil {
-		return 0, fmt.Errorf("cannot parse device size %q: %v", nospace, err)
+		return 0, fmt.Errorf("cannot parse blockdev size %q: %v", nospace, err)
 	}
 	return quantity.Size(sz), nil
 }
 
+func blockDeviceSizeInSectors(devpath string) (quantity.Size, error) {
+	// the size is always reported in 512-byte sectors, even if the device does
+	// not have a physical sector size of 512
+	// XXX: consider using /sys/block/<dev>/size directly
+	return blockdevCmd("--getsz", devpath)
+}
+
+func blockDeviceSectorSize(devpath string) (quantity.Size, error) {
+	// the size is reported in raw bytes
+	return blockdevCmd("--getss", devpath)
+}
+
 // onDiskVolumeFromPartitionTable takes an sfdisk dump partition table and returns
 // the partitioning information as an on-disk volume.
 func onDiskVolumeFromPartitionTable(ptable sfdiskPartitionTable) (*OnDiskVolume, error) {
```

Comment:
do we have it confirmed that sfdisk reports sizes in the multiples of logical sector size (actual one rather than fixed 512) in JSON format, even if it is old and only includes the sector size in the `--dump` output?

Reply:
Well I included the output of running `sfdisk --json` (but not with `--dump` though) from various systems, so I feel confident that this is the case, but perhaps you also want to see output with `--dump` too?

Reply:
Yes, for clarity it'd be great if you could attach the output of `--dump` too.

Reply:
ok, I started a new comment here with all the various command outputs on different devices: https://github.com/snapcore/snapd/pull/9895#issuecomment-785144419